// mongodb-memory-server-core/lib/util/resolveConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.envToBool = exports.envName = exports.resolveConfig = exports.processConfigOption = exports.findPackageJson = exports.setDefaultValue = exports.defaultValues = exports.DEFAULT_VERSION = exports.ENV_CONFIG_PREFIX = exports.ResolveConfigVariables = void 0;
const tslib_1 = require("tslib");
const camelcase_1 = (0, tslib_1.__importDefault)(require("camelcase"));
const new_find_package_json_1 = require("new-find-package-json");
const debug_1 = (0, tslib_1.__importDefault)(require("debug"));
const path = (0, tslib_1.__importStar)(require("path"));
const fs_1 = require("fs");
const utils_1 = require("./utils");
const log = (0, debug_1.default)('MongoMS:ResolveConfig');
/** Enum of all possible config options */
var ResolveConfigVariables;
(function (ResolveConfigVariables) {
    ResolveConfigVariables["DOWNLOAD_DIR"] = "DOWNLOAD_DIR";
    ResolveConfigVariables["PLATFORM"] = "PLATFORM";
    ResolveConfigVariables["ARCH"] = "ARCH";
    ResolveConfigVariables["VERSION"] = "VERSION";
    ResolveConfigVariables["DEBUG"] = "DEBUG";
    ResolveConfigVariables["DOWNLOAD_MIRROR"] = "DOWNLOAD_MIRROR";
    ResolveConfigVariables["DOWNLOAD_URL"] = "DOWNLOAD_URL";
    ResolveConfigVariables["PREFER_GLOBAL_PATH"] = "PREFER_GLOBAL_PATH";
    ResolveConfigVariables["DISABLE_POSTINSTALL"] = "DISABLE_POSTINSTALL";
    ResolveConfigVariables["SYSTEM_BINARY"] = "SYSTEM_BINARY";
    ResolveConfigVariables["MD5_CHECK"] = "MD5_CHECK";
    ResolveConfigVariables["ARCHIVE_NAME"] = "ARCHIVE_NAME";
    ResolveConfigVariables["RUNTIME_DOWNLOAD"] = "RUNTIME_DOWNLOAD";
    ResolveConfigVariables["USE_HTTP"] = "USE_HTTP";
    ResolveConfigVariables["SYSTEM_BINARY_VERSION_CHECK"] = "SYSTEM_BINARY_VERSION_CHECK";
    ResolveConfigVariables["USE_ARCHIVE_NAME_FOR_BINARY_NAME"] = "USE_ARCHIVE_NAME_FOR_BINARY_NAME";
})(ResolveConfigVariables = exports.ResolveConfigVariables || (exports.ResolveConfigVariables = {}));
/** The Prefix for Environmental values */
exports.ENV_CONFIG_PREFIX = 'MONGOMS_';
/** This Value exists here, because "defaultValues" can be changed with "setDefaultValue", but this property is constant */
exports.DEFAULT_VERSION = '5.0.13';
/** Default values for some config options that require explicit setting, it is constant so that the default values cannot be interfered with */
exports.defaultValues = new Map([
    // apply app-default values here
    [ResolveConfigVariables.VERSION, exports.DEFAULT_VERSION],
    [ResolveConfigVariables.PREFER_GLOBAL_PATH, 'true'],
    [ResolveConfigVariables.RUNTIME_DOWNLOAD, 'true'],
    [ResolveConfigVariables.USE_HTTP, 'false'],
    [ResolveConfigVariables.SYSTEM_BINARY_VERSION_CHECK, 'true'],
    [ResolveConfigVariables.USE_ARCHIVE_NAME_FOR_BINARY_NAME, 'false'],
]);
/**
 * Set an Default value for an specific key
 * Mostly only used internally (for the "global-x.x" packages)
 * @param key The Key the default value should be assigned to
 * @param value The Value what the default should be
 */
function setDefaultValue(key, value) {
    exports.defaultValues.set(key, value);
}
exports.setDefaultValue = setDefaultValue;
/** Cache the found package.json file */
let packagejson = undefined;
/**
 * Find the nearest package.json (that has an non-empty config field) for the provided directory
 * @param directory Set an custom directory to search the config in (default: process.cwd())
 * @returns what "packagejson" variable is
 */
function findPackageJson(directory) {
    var _a;
    for (const filename of (0, new_find_package_json_1.findSync)(directory || process.cwd())) {
        log(`findPackageJson: Found package.json at "${filename}"`);
        const readout = JSON.parse((0, fs_1.readFileSync)(filename).toString());
        /** Shorthand for the long path */
        const config = (_a = readout === null || readout === void 0 ? void 0 : readout.config) === null || _a === void 0 ? void 0 : _a.mongodbMemoryServer;
        if (!(0, utils_1.isNullOrUndefined)(config) && Object.keys(config !== null && config !== void 0 ? config : {}).length > 0) {
            log(`findPackageJson: Found package with non-empty config field at "${filename}"`);
            const filepath = path.dirname(filename);
            packagejson = {
                filePath: filepath,
                config: processConfigOption(config, filepath),
            };
            break;
        }
    }
    return packagejson;
}
exports.findPackageJson = findPackageJson;
/**
 * Apply Proccessing to input options (like resolving paths)
 * @param input The input to process
 * @param filepath The FilePath for the input to resolve relative paths to (needs to be a dirname and absolute)
 * @returns always returns a object
 */
function processConfigOption(input, filepath) {
    log('processConfigOption', input, filepath);
    if (typeof input !== 'object') {
        log('processConfigOptions: input was not a object');
        return {};
    }
    // cast because it was tested before that "input" is a object and the key can only be a string in a package.json
    const returnobj = input;
    // These are so that "camelCase" doesnt get executed much & de-duplicate code
    // "cc*" means "camelcase"
    const ccDownloadDir = (0, camelcase_1.default)(ResolveConfigVariables.DOWNLOAD_DIR);
    const ccSystemBinary = (0, camelcase_1.default)(ResolveConfigVariables.SYSTEM_BINARY);
    if (ccDownloadDir in returnobj) {
        returnobj[ccDownloadDir] = path.resolve(filepath, returnobj[ccDownloadDir]);
    }
    if (ccSystemBinary in returnobj) {
        returnobj[ccSystemBinary] = path.resolve(filepath, returnobj[ccSystemBinary]);
    }
    return returnobj;
}
exports.processConfigOption = processConfigOption;
/**
 * Resolve "variableName" value (process.env | packagejson | default | undefined)
 * @param variableName The variable to search an value for
 */
function resolveConfig(variableName) {
    var _a, _b, _c;
    return (_c = ((_b = (_a = process.env[envName(variableName)]) !== null && _a !== void 0 ? _a : packagejson === null || packagejson === void 0 ? void 0 : packagejson.config[(0, camelcase_1.default)(variableName)]) !== null && _b !== void 0 ? _b : exports.defaultValues.get(variableName))) === null || _c === void 0 ? void 0 : _c.toString();
}
exports.resolveConfig = resolveConfig;
exports.default = resolveConfig;
/**
 * Helper Function to add the prefix for "process.env[]"
 */
function envName(variableName) {
    return `${exports.ENV_CONFIG_PREFIX}${variableName}`;
}
exports.envName = envName;
/**
 * Convert "1, on, yes, true" to true (otherwise false)
 * @param env The String / Environment Variable to check
 */
function envToBool(env = '') {
    if (typeof env !== 'string') {
        log('envToBool: input was not a string!');
        return false;
    }
    return ['1', 'on', 'yes', 'true'].indexOf(env.toLowerCase()) !== -1;
}
exports.envToBool = envToBool;
// enable debug if "MONGOMS_DEBUG" is true
if (envToBool(resolveConfig(ResolveConfigVariables.DEBUG))) {
    debug_1.default.enable('MongoMS:*');
    log('Debug Mode Enabled, through Environment Variable');
}
// run this after env debug enable to be able to debug this function too
findPackageJson();
// enable debug if "config.mongodbMemoryServer.debug" is true
if (envToBool(resolveConfig(ResolveConfigVariables.DEBUG))) {
    debug_1.default.enable('MongoMS:*');
    log('Debug Mode Enabled, through package.json');
}
